sources/personio/helpers.py

```diff
@@ -0,0 +1,137 @@
+"""Personio source helpers"""
+from typing import Any, Iterable, Optional
+from urllib.parse import urljoin
+
+import backoff
+import inflection
+from dlt.common.time import ensure_pendulum_datetime
+from dlt.common.typing import Dict, TDataItems
+from dlt.sources.helpers import requests
+from requests.exceptions import RequestException
+
+
+def backoff_giveup(ex: RequestException) -> bool:
+    status_code = ex.response.status_code
+    if status_code >= 500 or status_code in [429]:
+        return False
+    return True
+
+
+class PersonioApi:
+    """A Personio API client."""
+
+    base_url = "https://api.personio.de/v1/"
+
+    def __init__(self, client_id: str, client_secret: str) -> None:
+        """
+        Args:
+            client_id: The client ID of your app.
+            client_secret: The client secret of your app.
+        """
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.access_token = self.get_token()
+
+    @backoff.on_exception(
+        backoff.expo,
+        RequestException,
+        max_tries=5,
+        giveup=backoff_giveup,  # type: ignore
+    )
+    def _request(
+        self,
+        method: str,
+        resource: str,
+        params: Optional[Dict[str, Any]] = None,
+        headers: Optional[Dict[str, Any]] = None,
+        json: Optional[Dict[str, Any]] = None,
+    ) -> Any:
+        """Make a request to Personio.
+
+        Args:
+            method: The HTTP method to use.
+            resource: The resource to get pages for (e.g. employees, absences, attendances).
+            params: Query params to include in the request.
+            headers: Headers to include in the request.
```

Comment:
the docstring is missing the json argument and (imo) it's better to use a different name, json is a library, conflicts may arise

Reply:
I will use the dlt request so it will not be used anymore.